### ISSUE TITLE
missing-argument false positive on super().__init__() when a Pydantic subclass overrides a required field with a default and defines an explicit __init__ (regression since 0.0.58, still present in 0.0.72)

### ISSUE THREAD

**Issue by Andrew-Tierno** [astral-sh/ty#4323]
## Summary

When a `pydantic.BaseModel` subclass:

1. overrides an ancestor's *required* field (declared via `Field(...)`, no
   default) with a field that **has** a default, and
2. defines its own explicit `__init__` that calls `super().__init__()` with
   no arguments,

`ty` reports a false `missing-argument` diagnostic on the `super().__init__()`
call, demanding the field the *ancestor* declares as required even though
the *subclass* (the actual runtime type being constructed) gives it a
default.

At runtime this is unambiguous: Pydantic's synthesized `BaseModel.__init__`
validates against the model's own (subclass) field definitions, not the
immediate ancestor's pre-override declaration. `object.__setattr__` is the
standard pattern for populating `PrivateAttr` fields from a custom
constructor, since `PrivateAttr` fields are exactly why these classes need a
hand-written `__init__` in the first place (to accept constructor arguments
that aren't validated Pydantic fields).

## Reproduction

`widget.py`:

```python
from pydantic import BaseModel, Field, PrivateAttr


class Tool(BaseModel):
    name: str = Field(...)


class Widget(Tool):
    name: str = "widget"

    _payload: str = PrivateAttr()

    def __init__(self, payload: str) -> None:
        super().__init__()
        object.__setattr__(self, "_payload", payload)


def make_widget() -> Widget:
    return Widget("value")


if __name__ == "__main__":
    widget = make_widget()
    assert widget.name == "widget"
    assert widget._payload == "value"
    print("runtime OK:", widget)
```

Commands:

```bash
mkdir ty-repro && cd ty-repro
# paste widget.py above into this directory, then:

uvx --with pydantic==2.13.2 python3 widget.py
# runtime OK: name='widget'

uvx --from 'ty==0.0.57' --with pydantic==2.13.2 ty check widget.py   # All checks passed!
uvx --from 'ty==0.0.58' --with pydantic==2.13.2 ty check widget.py   # error[missing-argument] — see below
uvx --from 'ty==0.0.70' --with pydantic==2.13.2 ty check widget.py   # still errors
uvx --from 'ty==0.0.72' --with pydantic==2.13.2 ty check widget.py   # still errors (0.0.72 is latest as of 2026-08-17)
```

## Expected vs actual

**Expected**: no diagnostic. `Widget("value")` — and therefore the
`super().__init__()` call inside `Widget.__init__` — is valid: `Widget.name`
has a default (`"widget"`), so `Tool`/`BaseModel`'s validated constructor for
the *actual* class being built requires nothing. This checked clean through
`ty` 0.0.57.

**Actual**:

```
error[missing-argument]: No argument provided for required parameter `name`
  --> widget.py:14:9
   |
14 |         super().__init__()
   |         ^^^^^^^^^^^^^^^^^^
   |

Found 1 diagnostic
```

The diagnostic always lands on the `super().__init__()` call itself (line 14
above), never on the external `Widget("value")` construction site.

## Environment

- ty versions tested: 0.0.57 (clean), 0.0.58 (first bad), 0.0.70, 0.0.72 (latest as of 2026-08-17, released 2026-08-14)
- Python: 3.13 (the `uvx` default interpreter in this repro); also confirmed on 3.12 and 3.14 in our internal monorepo
- pydantic: 2.13.2, via `uvx --with pydantic==2.13.2`

### Version

0.0.72 (2026-08-17)

**Comment by sharkdp** [comment]
Thank you for reporting this. A simplified version of your example is the following, which doesn't require `PrivateAttr`:

```py
from pydantic import BaseModel, Field

class Tool(BaseModel):
    name: str = Field(...)

class Widget(Tool):
    name: str = "widget"

    def __init__(self) -> None:
        super().__init__()  # error[missing-argument]: ... required parameter `name`
```

This is tricky for a type checker to model. In fact, mypy (with Pydantic plugin) and Pyrefly (with builtin Pydantic support) also report this as an error. At runtime, `Tool` has no `__init__` method, so `super().__init__()` actually calls `BaseModel.__init__`. However, to model dataclass-like behavior for Pydantic models, ty synthesizes a `Tool.__init__` method that requires the `name` field.

I'm not really sure what (if anything) to do about this. @carljm, you labeled this as a bug. Did you have a solution in mind?

**Comment by AlexWaygood** [comment]
> At runtime, `Tool` has no `__init__` method

Is that true for all pydantic `BaseModel` subclasses? Since we already have special casing in place for pydantic, could we just add special casing so that we understand that subclasses of `BaseModel` (unlike most dataclasses) do not synthesize `__init__` methods?

**Comment by sharkdp** [comment]
> could we just add special casing so that we understand that subclasses of `BaseModel` (unlike most dataclasses) do not synthesize `__init__` methods

But we *want* those synthesized `__init__` methods, so that we can validate explicit constructor calls. `Tool()` is and should be an error. Obviously, we don't *have* to model this using synthesized constructors, but that's the way it's currently implemented for all code generators.

**Comment by AlexWaygood** [comment]
Sorry, I should have read the issue text more closely. I see the issue. It's a bit strange that `super().__init__()` works at all, in that case? I agree that this seems hard to model with our current architecture.

**Comment by carljm** [comment]
Sorry, my triage was too shallow, I verified the discrepancy from runtime but did not think through the implementation consequences. I suppose we _can_ fix this by further special casing of Pydantic relative to other dataclass-likes. We could more closely model the runtime behavior by not synthesizing `__init__` methods on each model class, but rather special-casing the behavior of `BaseModel.__init__` when that is the resolved initializer, so that it builds a constructor signature based on the fields of the actual type being constructed. But I'm not sure that we _should_ do that; it seems like a lot of dedicated machinery to fix a fairly narrow limitation currently shared by all type checkers.

I'm not opposed to closing this issue, or to leaving it open (but out of any milestone) on the theory that in principle it is a bug (we fail to model pydantic accurately) and we wouldn't be opposed to fixing it in theory, but it may never be worth the implementation complexity required.
